Epub accessibility check

mwbenowitz opened the pull request:
This adds a new step (and tests) to the s3 storage process that generates an accessibility report for all ePubs being stored. This report is generated through the Daisy Consortium's Ace Report tool, which is running behind an API on an ec2 instance.

There are various reasons why this is running on a standalone instance, the main being that Ace includes Chromium as a dependency and it is difficult/inadvisable to run a Lambda which includes chromium. 

This branch adds a new output object to the Kinesis stream that includes the full Report data and summaries, this is to be read by the database updater add to add to a new table in the ResearchNow data model.

mwbenowitz committed:
SFR-193 Added Accessibility Scoring using Daisy/Ace

This generates a score and accessibility report using the Ace
library from Daisy. This contains basic metadata about the report
as well as the full JSON block produced for future reference.

The report takes some time to produce, though that may be to local
limitations, it should be reviewed for performance in the dev
environment.

This adds a new record type to the output Kinesis stream, as these
records should populate a new table in the db that relates to both
items (the score is calculated from the zipped version, but should
apply to both)

mwbenowitz committed:
SFR-193 Completed Accessibility Score Test Coverage

mwbenowitz committed:
SFR-193 Forgot to uncomment a test

mwbenowitz committed:
Merge branch 'development' into epub-accessibility-check

mwbenowitz committed:
SFR-193 Fix failing tests & ensure all promises are resolved

mwbenowitz committed:
SFR-193 Initial Travis Deployment

mwbenowitz committed:
Updated travis vars

mwbenowitz committed:
SFR-193 Move Accessibility Checker to standalone API

Due to conflicting requirements of the Ace Report generator and the
AWS Lambda environment, the actual work of generating those reports
has been spun off into a separate API running on a t2.micro ec2
instance. (These issues primarily relate to Ace's reliance on Chromium,
which puts demands on package size, startup time and dependency
resolution in the Lambda environment)

mwbenowitz committed:
SFR-193 Fix environment files and better handled errors

mwbenowitz committed:
SFR-193 Added documentation explaining link to Ace report generator
service

mwbenowitz committed:
Fixed travis config (hopefully to trigger a build)

mwbenowitz committed:
Fix Test Failures for Accessibility Report

This package's test suite was not updated after the Accessibility
report module was split into a standalone API. These unecessary
tests were causing builds to fail. They have been re-written along
with a few other minor errors discovered during testing.

Finally an unused (and deprecated) package -- zlib -- was removed

gkallenberg commented on index.js:
```diff
@@ -1,71 +1,109 @@
+import Parser from './src/epubParsers'
+import ResHandler from './src/responseHandlers'
 
 const fileNameRegex = /[0-9]+[.]{1}epub[.]{1}(?:no|)images/
```
wondering if the `[.]` should be escaped as in `[\.]` if the period or dot is what is being tested.

mwbenowitz replied:
within the brackets escaping is unnecessary. I can add it in if you think its clearer

gkallenberg commented on src/epubParsers.js:
```diff
@@ -1,113 +1,110 @@
-    region: 'us-east-1',
-    logger: process.stdout
+  region: 'us-east-1',
+  logger: process.stdout
```
we'll need to add a logger like winston for sending standard log messages to CloudWatch

mwbenowitz replied:
I've added Winston as the logger for the function but not on this line. I think that this output should be reflected in CloudWatch, but I need to look at how best to integrate winston with the AWS SDK. There is at least one library that does so, but I haven't reviewed it yet

gkallenberg commented on test/accessibilityCheck.test.js:
```diff
@@ -0,0 +1,46 @@
+      expect(resp.data).to.equal('Some Test Data')
+    })
+
+    it('should get a throw an error if it does not get a report', async () => {
```
edit: `should throw an error...`

mwbenowitz replied:
thanks!

gkallenberg commented on test/handler.test.js:
```diff
@@ -0,0 +1,179 @@
+      parseStub.restore()
+    })
+
+    it('should return 500 if no records in array', () => {
```
Is an empty array indicating no records were found? It's a valid array object, but without a set of records to process to processing should end. I think we talked about this in stand up -- using a different status code like a 204 No Content response? That would indicate that the request was fulfilled and terminate the process.

gkallenberg reviewed (commented):
Looks good. I would add some of the basics from our [documentation standards](https://github.com/NYPL/engineering-general/blob/master/standards/documentation.md) to the README like requirements, dependencies, installation instructions, testing instructions. I need to do this on my repos as well. Let me know if the comments make sense or not.

Check your linter errors.

And we should have a standard logging package like `winston` to handle errors in a standard way for CloudWatch.

Should also talk about retry scenarios for 5xx responses since servers/services may be unavailable temporarily.

seanredmond commented on index.js:
```diff
@@ -1,71 +1,109 @@
+      "message": "Created Accessibility Score",
+      "data": accessReport
+    }
+  } catch(err) {
```
I wonder if it's cleaner just to rethrow the error here? And similarly in the filename match below and other places, there are a bunch of spots where a 500 error might be returned, would it be simpler to throw errors and have them caught, logged, and converted to a return code in one place?

mwbenowitz replied:
After doing the refactoring suggested below, this is essentially what it does now. It throws Lambda Errors, which are handled/logged at a higher level

seanredmond commented on index.js:
```diff
@@ -1,71 +1,109 @@
-                }
-                resultHandler(handleResp)
-            })
+exports.parseRecord = (record) => {
```
This feels like two functions to me. One to parse the JSON, check the filename, and either throw an error or return the filename, id, and date; and another to take those and do the existence checking and exploding.

mwbenowitz replied:
I agree, I've split out the functionality into two separate methods. It revealed a few issues with the code and made the test suite coverage much more accurate

mwbenowitz committed:
Updated Readme with dependencies, deployment and testing

mwbenowitz committed:
Added winston as a logging service for the function

mwbenowitz committed:
Updated with various updates to improve functionality

This includes splitting several functions to better
split up responsibility as well as provide better
logging, among other improvements

mwbenowitz commented:
Sorry for the delay on responding to these comments! Thanks for taking a look at everything. In addition to the specific comments here I also updated the Readme to have more details from the engineering guidelines and addressed the linter errors as Greg pointed out. I'm merging these changes now because I'm updating the input/response formats to better match the data model that we've now established with the persistence layer and it will make it easier to integrate the response from this function (especially for the reports)
